### ISSUE TITLE
as_draws_rvars() performance with large-but-not-ridiculous input

### ISSUE THREAD

**Issue by mike-lawrence** [stan-dev/posterior#119]
So I have a stan file to generate data:
```stan
// mvn_generate.stan
data{
	// n: number of subj
	int<lower=1> n ;
	// k: number of variables modelled as intercorrelated
	int<lower=1> k ;
}
parameters{
	real dummy_ ;
}
generated quantities{
	// means: mean (across subj) for each coefficient
	vector[k] means ;
	for(i_k in 1:k){
		means[i_k] = std_normal_rng() ;
	}
	// sds: sd (across subj) for each coefficient
	vector[k] sds ;
	for(i_k in 1:k){
		sds[i_k] = weibull_rng(2,1) ;
	}
	// chol_corr: population-level correlations (on cholesky factor scale) amongst within-subject predictors
	matrix[k,k] chol_corr = lkj_corr_cholesky_rng(k,1) ;
	// y: array of values for each subj & coef
	vector[k] y[n] ;
	for(i_n in 1:n){
		y[i_n] = multi_normal_cholesky_rng(
			means
			, diag_pre_multiply(sds,chol_corr)
		) ;
	}
}
```

And I use it to generate quantities via cmdstanr:
```r
n = 4e2
k = 4e2
mod = cmdstanr::cmdstan_model('mvn_generate.stan')
system.time(
	quantities <- mod$generate_quantities(
		data = lst(n,k)
		, fitted_params = posterior::as_draws_array(tibble(dummy_=0))
	)
) #about .5s
file.size(quantities$output_files()) #about 8MB
system.time(y_draws <- quantities$draws('y') ) #about 6s
object.size(y_draws) #about 13MB
system.time(y_rvars <- posterior::as_draws_rvars(y_draws) ) #about 90s
```
As you can see, it takes a minute and a half to run `as_draws_rvars()`. Is this expected performance? 

As I increase `n` or `k` the time that cmdstan takes to generate the samples and the time that it takes to collect the draws seems to scale reasonably, but the call to `as_draws_rvars()` seems to increase dramatically. (Note I know this isn't the intended use case for rvars, where there is only one draw, but thought posting this reprex might be helpful)

**Comment by mjskay** [comment]
Thanks for this! I did some profiling and narrowed it down to this line:

https://github.com/stan-dev/posterior/blob/0184c56f0f6f7a7f9183f418983c1d028fef2942/R/as_draws_rvars.R#L72-L73

If I change `do_call` to `do.call` the runtime goes from 30 seconds to 1 second on my machine. On the one hand, that's an easy fix. On the other hand, this suggests maybe some issues with `do_call()`? The profiler is not showing me much useful information about what's happening inside it to cause things to be slow, I think maybe the final `eval()` is the culprit. Possibly it is a result of the large number of arguments passed (the roundtrip via a string and back to an expression might very slow for a lot of args --- perhaps a version that builds an expression directly without needing to hit the parser would do better), but I'm not sure. @paul-buerkner?

**Comment by paul-buerkner** [comment]
interesting. thank you! you can change to do.call for now and until we find
out if/how we can improve do_call. for context I wrote the do_call function
to avoid deparsing arguments which may cause trouble in nested do.calls


Matthew Kay ***@***.***> schrieb am Sa., 3. Apr. 2021, 08:27:

**Comment by paul-buerkner** [comment]
I have decided to replace do_call with do.call again in all of posterior. Until this issue occured, I had thought do_call was just a better version of do.call without drawbacks but this is not the case. The problem do.call has when being used in a nested manner occurred in brms and I don't see a place in the current posterior codebase where this is may happen, nor do I see that it is likely to happen in the future of posterior. Hence, I think it is safe to just replace. 

Closing this issue.
